### ISSUE TITLE
When are stack traces set on `Error`s.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#3627]
TL;DR: We should introduce some leniency in how and when `.stackTrace` is set, because the current specification doesn't match implementations, and they deviated with good reason. (And nobody noticed.)

The language requires that an object extending `Error` from `dart:core` gets its `.stackTrace` filled in when it's thrown the first time.

Some implementations have bugs where the `.stackTrace` is not `null` until the first throw. Those are easily fixed, but while fixing that, I noticed another, less clearly wrong, divergence between platforms.

As specified, when executing `throw e`, if `e` evaluates to an object whose runtime type extends `Error` which has not been thrown before, the *current stack trace* (the same one that is thrown along with the object) is also set on the error object.

> ```latex
> If $v$ is an instance of class \code{Error} or a subclass thereof,
> and it is the first time that \code{Error} object is thrown,
> the stack trace $t$ is stored on $v$ so that it will be returned
> by the \code{stackTrace} getter inherited from \code{Error}.
> 
> \commentary{%
> If the same \code{Error} object is thrown more than once,
> its \code{stackTrace} getter will return the stack trace from
> the \emph{first} time it was thrown.%
> }
> ```

That means that the `.stackTrace` is, as specified, set at the `throw`, not at the `catch`.
However, JS compilers seem to implement this using the JS capturing of stack traces at throws, which means that they won't have access to that stack trace until the JS throw is caught. (At that point they store the JS error on the Dart object, so the Dart object can always look up the JS stack trace.)

But it means that the an intermediate *finally* can see the error in a non-filled-in state, and if a `finally` control-flows its way out of the throw, the error is never caught and the stack trace never set.

We probably want to *accept that behavior*, because otherwise it becomes extremely hard to implement Dart throwing in JavaScript. You'll need a double-throw, one to get the current stack trace, store it on the error, and then another to throw it for real. That's expensive and verbose, so not really an option.

## Proposal
We change the specification of `throw` to say:

> If *v* is an instance of a subclass of the class `Error`,
> then the current stack trace may be stored on the *v* object,
> making it accessible as a `StackTrace` object via the `stackTrace` getter.
> The stack trace may be stored at any time from the execution
> of the `throw` that throws *v* with that stack trace,
> until a `catch` clause catches the thrown object.
> _Any such catch clause will definitely have access to a stack trace._
> The current stack trace will not be stored if another stack trace 
> has already been stored on the same object.
> The stack trace may or may not be stored if the thrown object
> never reaches a `catch` clause. _That can happen if an intermediate
> `finallly` clause exits with a different control flow, or if the thrown
> object is caught by a `try`/`on` clause with no `catch` clause._

That is, the stack trace may be set immediately, or it may be set at the first `catch` catching the error.
I leave it open whether an `on Error { }` will definitely set the stack or not. It probably should, and it's probably not a problem to require it, but we should make safe first.

Nobody should ever *care* when the stack trace is set, as long as it's set before the error is caught.
If the error isn't caught at all, it's very likely to never be looked at again. 
In general, nobody hangs on to an `Error` after throwing it. The first time anyone looks at it again is when it's caught.

What can, and does, happen is that an error is caught and *rethrown*, and our current behavior of keeping the original stack trace is probably optimal for that. (The only alternative would be to capture multiple stack traces and concatenating them.)
